File: Old Code/WurmRancher/WurmRancher/GrassEater.cs
using System;
using System.Net;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Documents;
using System.Windows.Ink;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Shapes;
using System.Windows.Media.Imaging;
using Microsoft.Xna.Framework.Audio;

namespace WurmRancher
{
    public class GrassEater : LaserDestructablePiece
    {
        const int height = 30, width = 30;
        
 
        static BitmapImage GrassEaterBitmap = new BitmapImage();
        static SoundEffect GrassEaterDieSoud = SoundEffect.FromStream(WRResources.gruntz);

        static GrassEater()
        {
            GrassEaterBitmap.SetSource(WRResources.grass_eater);
           
        }

        public GrassEater(GameControl theControl_)
            : base(theControl_,height, width, Timing.RelativeSpeedToPixelsPerFrame(Timing.relGrassEaterSpeed), Timing.RelativeRotateToRadiansPerFrame(Timing.relGrassEaterRotate))
        {
            PieceImage.Source = GrassEaterBitmap;
        }

        protected override SoundEffect DieSound
        {
            get
            {
                return GrassEaterDieSoud;
            }
        }

        GoodGrass target_plant = null;
        public override void Update()
        {
            if (target_plant != null && Util.Distance(this, target_plant) < 1 && !(this.hit))
            {
                target_plant.Eat();
                if (target_plant.Eaten)
                    target_plant = null;
            }

            if (target_plant == null && this.resting) // find a new destination!
            {
                target_plant = theControl.GetClosestPlant<GoodGrass>(this);
                if (target_plant != null)
                {
                    this.SetDestination(target_plant.CenterPoint);
                    target_plant.Dibs(10);
                }
                else
                    this.SetDestination(this.RandomPointOnField());
            }
            base.Update();
        }
    }
}

File: ts/grassEater.ts
import { EdiblePlant } from "./ediblePlant.js";
import { DistanceObjects, SetTargetPlant } from "./gameControl.js";
import { LaserDestructablePiece } from "./laserDestructablePiece.js";
import { GrassChaser } from "./predPrey.js";
import { GrassEaterDeathSound, grassEaterImage } from "./resources.js";
import { GrassEaterRotate, GrassEaterSpeed } from "./timing.js";

const height = 30;
const width =30;


export class GrassEater extends LaserDestructablePiece implements GrassChaser{
    Layer = 6;
    Name = "GrassEater";
    hit = false;
    targetPlant :EdiblePlant;

    constructor(){
        super(height,width, GrassEaterSpeed, GrassEaterRotate);
        this.PieceImage = grassEaterImage;
        this.LaserHitSound = GrassEaterDeathSound;
    }

    Update(time_step:number) : void{
        if (this.hit){
            super.Update(time_step);
            return;
        }
        
        if (this.targetPlant && DistanceObjects(this, this.targetPlant) < 1 && !(this.hit))
        {
            this.targetPlant.Eat(time_step);
            if (this.targetPlant.Eaten)
                this.targetPlant = null;
        }

        if (this.targetPlant == null && this.resting) // find a new destination!
            SetTargetPlant(this, ["GoodGrass"]);

        //console.log("about to call super",this);
        super.Update(time_step);
    }
    
    PreyLost(): void {
        if (this.targetPlant){
            this.targetPlant.chaser = null;
            this.targetPlant = null;
        }
        this.resting = true;
    }

    Hit():void{
        this.PreyLost();
    }

}